src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/TransportCreation/NamedPipeTransportCreationHandler.cs

```diff
@@ -0,0 +1,37 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Data.SqlClient.SNI;
+
+namespace Microsoft.Data.SqlClientX.Handlers.TransportCreation
+{
+    internal sealed class NamedPipeTransportCreationHandler : ReturningHandler<ConnectionHandlerContext, Stream>
+    {
+        /// <inheritdoc />
+        public override async ValueTask<Stream> Handle(ConnectionHandlerContext parameters, bool isAsync, CancellationToken ct)
+        {
+            Debug.Assert(parameters is not null, "Parameters must not be null");
+            Debug.Assert(parameters.DataSource is not null, "Data source must not be null");
+
+            // This handler can only handle connection attempts if named pipes are included
+            if (parameters.DataSource.ResolvedProtocol is not (DataSource.Protocol.NP or DataSource.Protocol.None))
+            {
+                return await HandleNext(parameters, isAsync, ct);
+            }
+            
+            if (parameters.DataSource.ResolvedProtocol is DataSource.Protocol.None)
+            {
+                // @TODO: Support named pipe connections
+                return await HandleNext(parameters, isAsync, ct);
```

Comment:
Help me understand the `ReturningHandler` The NP handler ends up as the last handler in the returning handler chain. 
In that case, there would be no `HandleNext` call right? Either it connects or throws ?

Reply:
The idea was effectively for the handlers to not know where they are in the chain or if any other handlers are in the chain. So, if it can't handle it, it passes it onto the next handler - the `HandleNext` method throws if it's at the end of the chain.

I think I made a slight mistake in the NamedPipe implementation by explicitly throwing `NoSuitableHandlerFoundException` when in that case it should be throwing `NotImplementedException`. In fixing the previous comment (about inverting the condition), I also fixed this. Hopefully it makes more sense now.